1D Ocean Column MARBL Interface + Obs Converter

mgharamti opened the pull request:
## Description:
<!--- Describe your changes -->
A new model interface to the biogeochemistry model MARBL. The work done here is contributed by [Robin Armstrong](https://www.cam.cornell.edu/research/grad-students/robin-armstrong). Thanks Robin!

### Fixes issue
This address all comments, suggestions, and questions raised on the original PR [MARBL-DART #701](https://github.com/NCAR/DART/pull/701). It also fixes issue #614 

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

### Documentation changes needed?
<!-- Put an `x` in all the boxes that apply: -->
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.

### Tests
Ran several DART program to assess the behavior: `model_mod_check`, `filter`, and `bats_to_obs`
My tests are here: 
- Model Interface: `/glade/work/gharamti/DART/bgc_work/models/MARBL_column/work`
- Obs Converter: `/glade/work/gharamti/DART/bgc_work/observations/obs_converters/BATS`

## Checklist for merging

- [ ] Updated changelog entry
- [x] Documentation updated
- [ ] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [x] Dataset needed for testing available upon request
- [x] Dataset download instructions included
- [ ] No dataset needed

hkershaw-brown reviewed (changes requested):
Hi Moha, 
Thanks for this. I haven't looked at the code yet. 
There are some warnings in the docs build. 

```
(py-docs) [hkershaw:DOCS]() > make html
Running Sphinx v6.2.1
/Users/hkershaw/DART/DOCS/py-docs/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(
building [mo]: targets for 0 po files that are out of date
writing output... 
building [html]: targets for 237 source files that are out of date
updating environment: [new config] 237 added, 0 changed, 0 removed
reading sources... [100%] theory/readme                                                                                                                       
/Users/hkershaw/DART/pull_requests/pull_726/models/MARBL_column/readme.rst:28: ERROR: Unexpected indentation.
/Users/hkershaw/DART/pull_requests/pull_726/models/MARBL_column/readme.rst:34: ERROR: Unexpected indentation.
/Users/hkershaw/DART/pull_requests/pull_726/models/MARBL_column/readme.rst:45: ERROR: Unexpected indentation.
/Users/hkershaw/DART/pull_requests/pull_726/models/MARBL_column/readme.rst:46: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/hkershaw/DART/pull_requests/pull_726/observations/obs_converters/BATS/readme.rst:20: ERROR: Unexpected indentation.
/Users/hkershaw/DART/pull_requests/pull_726/observations/obs_converters/BATS/readme.rst:56: ERROR: Unexpected indentation.
/Users/hkershaw/DART/pull_requests/pull_726/observations/obs_converters/BATS/readme.rst:79: WARNING: Inline literal start-string without end-string.
/Users/hkershaw/DART/pull_requests/pull_726/observations/obs_converters/BATS/readme.rst:97: WARNING: Title underline too short.

Climatology:
~~~~~~~~~~~
looking for now-outdated files... none found
pickling environment... done
checking consistency... /Users/hkershaw/DART/pull_requests/pull_726/models/MARBL_column/readme.rst: WARNING: document isn't included in any toctree
/Users/hkershaw/DART/pull_requests/pull_726/observations/obs_converters/BATS/readme.rst: WARNING: document isn't included in any toctree
done
preparing documents... done
writing output... [100%] theory/readme                                                                                                                        
generating indices... genindex done
writing additional pages... search done
copying images... [100%] guide/images/WRF_tutorial_evolution2.png                                                                                             
copying downloadable files... [100%] ../DOCS/guide/DART_LAB/presentation/DART_LAB_Section05.pdf                                                               
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 10 warnings.

The HTML pages are in dart-docs/html.
```
https://readthedocs.org/api/v2/build/25474391.txt

In addition to adding to the toc:

The converter should be added to 'converter programs' https://docs.dart.ucar.edu/en/latest/observations/obs_converters/README.html#converter-programs

The model should be added to 'supported models'
https://docs.dart.ucar.edu/en/latest/models/README.html

hkershaw-brown commented:
a question/note on the parameter estimation only. 

This is the docs:

"Parameters Estimation only: where only the parameters are constrained using the data. DART will still need to read in the state to construct ensemble covariances and compute innovations. The only difference is that the ensemble increments are only regressed onto the unknown parameters. To achive this goal, you’ll need to set estimate_params = .true. and turn the update status for all state variable to NO_COPY_BACK

This ensures that the state will not be updated. The NO_COPY_BACK option is added as the 5th entry in the state table (after the variable name, its associated quantity and its physical bounds) within &model_nml"

NO_COPY_BACK means the state is not written out to a restart file. The state is still updated by the assimilation, and inflated.

hkershaw-brown commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+
+call get_model_variable_indices(index_in, lon_index, lat_index, level, kind_index=local_qty)
+
+! Make sure the longitude is properly assigned
+geolon = station_location(1)
+geolat = station_location(2)
+
+if (geolon < 0.0_r8) geolon = 360.0 + geolon
```
The station_location is constant, so you only have to this  code once in static_init_model.
The model is tiny, so performance wise this does not make a difference, but it might be clearer to have static things in static_model_init.

mgharamti replied:
good suggestion!

hkershaw-brown commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+integer :: param_dom_id  ! used to access MARBL internal parameters
+integer :: nfields ! number of fields in the state or parameter vector
+integer :: nz      ! the number of vertical layers
+integer :: model_size
```
model_size is an i8 integer in dart.
```suggestion
integer(i8) :: model_size
```

mgharamti replied:
Done

hkershaw-brown commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+end do
+
+
+! performing the interpolation for each ensemble member individually.
```
This is a question/confirmation for my own understanding. There is no horizontal interpolation, in MARBL_column.
All four gridpoints have identical values for a given level so you only do vertical interpolation, and you just pick point (1,1).  You could pick (2,2) and the interpolation would give the same answer.

mgharamti replied:
Yes, you're right. All four corners are identical columns so technically you can select any corner. This also means that no horizontal interpolation is needed.

hkershaw-brown commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+integer, parameter :: modelparams_table_width = 5
+
+! defining the variables that will be read from the namelist
+character(len=256) :: template_file(2)
```
Because you have this namelist as 2 values, you're requiring that people give two filenames. So even if they 

If you only give one file,  but estimate_params       = .true.,  you have  you get errors like this: 

```
***************** RUNNING    TEST 0    ***********************
 -- Reading the model_mod namelist and implicitly running static_init_model
**************************************************************
 
 Assimilate_these_obs_types:
    BATS_OXYGEN
    BATS_ALKALINITY
    BATS_INORGANIC_CARBON
    BATS_SILICATE
    BATS_PHOSPHATE
    BATS_NITRATE
 Evaluate_these_obs_types:
    BATS_ORGANIC_CARBON
    BATS_NITROGEN
 Use the precomputed Prior Forward Operators for these obs types:
    none
 
 ERROR FROM:
  source : netcdf_utilities_mod.f90
  routine: load_state_variable_info nf90_open
  message: 
 
 
 : NetCDF: Malformed URL
  message: ...
```

It would be better to have two namelist options, one for each of state, parameter:
state_template_file = 'state.nc'
parameter_template_file = 'param.nc

In the same way that you hae separate namelist entries for 'model_state_variables' and 'model_parameters'

mgharamti replied:
Done

hkershaw-brown commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+
+! defining the variables that will be read from the namelist
+character(len=256) :: template_file(2)
+character(len=256) :: ocean_geometry
+real(r8)           :: station_location(2)
+integer            :: time_step_days
+integer            :: time_step_seconds
+logical            :: estimate_params
```
all the namelist variables need default values.  

otherwise you have the value set to whatever junk is in memory if the namelist value is not set in input.nml

hkershaw-brown reviewed (changes requested):
Next batch of review comments for model_mod.

Couple of definite changes:
* namelist items need default values
* param file and state file template files, since you have separate namelist options for turning these on and off, I think it would be better to have a param_template_file and a state_template_file
* model_size is i8 in dart
* for get_close_obs|state this is just passing through to the location mod, so you can use the location mod code directly. 

Alphabetical order / style notes:
* MARBL_column before MOM6 https://dart-documentation--726.org.readthedocs.build/en/726/models/README.html
* parameters are CAPITAL LETTERS
* remove commented out print statements



A question on the interpolation, there is no horizontal interpolation so just confirming that that is correct?

The model_time is set for days from 1982. Is this true for all MARBL runs? 

There are some other bits and bobs of suggestions for unused variables and old print statements.

hkershaw-brown reviewed (changes requested):
see comment before.

mgharamti commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+
+module model_mod
+
+! This is a template showing the interfaces required for a model to be compliant
+! with the DART data assimilation infrastructure. Do not change the arguments
+! for the public routines.
```
Done

mgharamti commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+real(r8) :: basin_depth(2,2)
+
+! parameters to be used in specifying the DART internal state
+integer, parameter :: modelvar_table_height = 13
+integer, parameter :: modelvar_table_width = 5
+integer, parameter :: modelparams_table_height = 1
+integer, parameter :: modelparams_table_width = 5
```
Done

mgharamti commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+
+!------------------------------------------------------------------
+!
+! Called to do one time initialization of the model. As examples,
+! might define information about the model size or model timestep.
+! In models that require pre-computed static data, for instance
+! spherical harmonic weights, these would also be computed here.
```
Done

mgharamti commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+
+! performing the interpolation for each ensemble member individually.
+do ens_index = 1, ens_size
+    !print *, "BASIN DEPTH     = ",-basin_depth(1,1)
+    !print *, "----------------------------------------------------"
+    !print *, "computing centers of layers"
+    !print *, "----------------------------------------------------"
```
Done

mgharamti commented on models/MARBL_column/readme.rst:
```diff
@@ -0,0 +1,133 @@
++-------------------------------------+--------------------+------------------------------------------------------------+
+| Item                                | Type               | Description                                                |
++=====================================+====================+============================================================+
+| ``template_file``                   | character(len=256) | MARBL restart file including MARBL's prognostic variables  |
```
Done

mgharamti commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+                                get_varid_from_kind, get_dart_vector_index, &
+                                get_num_domains
+
+use obs_kind_mod, only : get_index_for_quantity, QTY_U_CURRENT_COMPONENT, &
+                         QTY_V_CURRENT_COMPONENT, QTY_DRY_LAND, &
+                         QTY_LAYER_THICKNESS
```
Done

mgharamti commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+
+integer :: ncid
+
+character(len=*), parameter :: routine = 'read_num_layers'
```
Done

mgharamti commented on models/MARBL_column/model_mod.f90:
```diff
@@ -0,0 +1,638 @@
+type(time_type) :: mom6_time
+integer :: mom_base_date_in_days, mom_days
+
+mom_base_date_in_days = 139157 ! 1982 1 1 0 0
```
Yes, I believe this the time format for MARBL. I think it has to do with dates related to obs availability. I recall Robin spending some time on this before figuring out the correct format.

mgharamti commented:
Helen, thanks for the useful comments. I think I addressed all your comments/suggestions and incorporated them in the code. Thanks for clarifying the no-copy-back option. I've edited the text in the documentation to make it more clear. Please let me know if you have any questions.

hkershaw-brown commented:
cool thanks Moha, I'll move on to the obs converter

hkershaw-brown commented on observations/obs_converters/BATS/bats_to_obs.f90:
```diff
@@ -0,0 +1,435 @@
+                              do_nml_file, do_nml_term
+use  time_manager_mod, only : time_type, set_calendar_type, set_date, &
+                              operator(>=), increment_time, get_time, &
+                              operator(-), NOLEAP, GREGORIAN, operator(+), &
```
NOLEAP not used

hkershaw-brown commented on observations/obs_converters/BATS/bats_to_clim_obs.f90:
```diff
@@ -0,0 +1,368 @@
+                              do_nml_file, do_nml_term
+use  time_manager_mod, only : time_type, set_calendar_type, set_date, &
+                              operator(>=), increment_time, get_time, &
+                              operator(-), NOLEAP, GREGORIAN, operator(+), &
```
NOLEAP not used

hkershaw-brown commented on observations/obs_converters/BATS/bats_to_obs.f90:
```diff
@@ -0,0 +1,435 @@
+
+program bats_to_obs
+
+use         types_mod, only : r8, PI, DEG2RAD
```
PI, DEG2RAD not used

hkershaw-brown commented on observations/obs_converters/BATS/bats_to_obs.f90:
```diff
@@ -0,0 +1,435 @@
+
+      ! unit corrections from BATS to MARBL
+      if((OTYPE_ORDERING(otype_index) == BATS_ALKALINITY) .or. (OTYPE_ORDERING(otype_index) == BATS_INORGANIC_CARBON)) then
+         ovalue = ovalue*1.026
```
just a comment, 1.026 is a wild unit conversion.

mgharamti replied:
agreed :)

hkershaw-brown commented on observations/obs_converters/BATS/bats_climatology.py:
```diff
@@ -0,0 +1,156 @@
+
+# setting up a CSV file to write the climatology into; this will be read by the DART data converter.
+
+os.system("rm -f "+output_txt_path)
+output_txt = open(output_txt_path, "w")
```
this rm -f  seems dangerous.

Doesn't open(output_txt_path, "w") overwrite the file if it exists?

mgharamti replied:
Agreed

hkershaw-brown commented on observations/obs_converters/BATS/bats_climatology.py:
```diff
@@ -0,0 +1,156 @@
+
+# setting up a netCDF file to write the climatology into; this will be useful for creating MARBL-DART diagnostics.
+
+os.system("rm -f "+output_nc_path)
+output_nc = nc.Dataset(output_nc_path, "w")
```
same rm -f comment

hkershaw-brown commented on observations/obs_converters/BATS/bats_climatology.py:
```diff
@@ -0,0 +1,156 @@
+bats_data   = open(input_file_path, "r")
+line_number = 0
+
+for line in bats_data.readlines():
```
Not sure how much python reviewing you want on this,

with open is recommended rather than open and close.
You can slice with readlines to skip the first nlines. 

```
with open(input_file_path, 'r') as bats_data:
    for line in bats_data.readlines()[first_data_line:]
```

mgharamti replied:
I like with open more. The code will be modified accordingly

hkershaw-brown commented on observations/obs_converters/BATS/bats_climatology.py:
```diff
@@ -0,0 +1,156 @@
+import os
```
#!/usr/bin/env python3

hkershaw-brown reviewed (changes requested):
next batch of comments on obs_converters.

hkershaw-brown commented on assimilation_code/modules/observations/default_quantities_mod.f90:
```diff
@@ -382,6 +387,23 @@
+!   QTY_DISSOLVED_ORGANIC_P
+!   QTY_DISSOLVED_INORGANIC_IRON
+!   QTY_SURFACE_CHLOROPHYLL
+!   QTY_LAYER_THICKNESS
```
edit: there are a few duplicates in here: 

```
Duplicate lines containing 'QTY' found:
!   QTY_ALKALINITY (appears 2 times)
!   QTY_DISSOLVED_INORGANIC_CARBON (appears 2 times)
!   QTY_DISSOLVED_INORGANIC_IRON (appears 2 times)
!   QTY_DISSOLVED_INORGANIC_SIO3 (appears 2 times)
!   QTY_DISSOLVED_ORGANIC_CARBON (appears 2 times)
!   QTY_DISSOLVED_ORGANIC_NITROGEN (appears 2 times)
!   QTY_DISSOLVED_ORGANIC_P (appears 2 times)
!   QTY_DISSOLVED_OXYGEN (appears 2 times)
!   QTY_LAYER_THICKNESS (appears 2 times)
!   QTY_MESOZOOPLANKTON_CARBON (appears 2 times)
!   QTY_MICROZOOPLANKTON_CARBON (appears 2 times)
!   QTY_NITRATE_CONCENTRATION (appears 2 times)
!   QTY_PHOSPHATE_CONCENTRATION (appears 2 times)
!   QTY_PHYTOPLANKTON_BIOMASS (appears 2 times)
!   QTY_SURFACE_CHLOROPHYLL (appears 2 times)
```

hkershaw-brown commented:
@hkershaw-brown todo: test netcdf IO change

mgharamti commented on observations/obs_converters/BATS/bats_to_clim_obs.f90:
```diff
@@ -0,0 +1,368 @@
+!     adapted for more generic use 11 Mar 2010, nancy collins, ncar/image
+!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+subroutine create_3d_obs(lat, lon, vval, vkind, obsv, otype, oerr, day, sec, qc, obs)
```
I don't think there is any special reason for that, I've removed the ones in the converter and used the tools in obs_utilities_mod and got the same answer. So, I'm cleaning this up as recommended.

mgharamti commented on observations/obs_converters/BATS/bats_to_obs.f90:
```diff
@@ -0,0 +1,435 @@
+!     adapted for more generic use 11 Mar 2010, nancy collins, ncar/image
+!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+subroutine create_3d_obs(lat, lon, vval, vkind, obsv, otype, oerr, day, sec, qc, obs)
```
Done

mgharamti commented on observations/obs_converters/BATS/bats_to_obs.f90:
```diff
@@ -0,0 +1,435 @@
+!     created Mar 8, 2010   nancy collins, ncar/image
+!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+subroutine add_obs_to_seq(seq, obs, obs_time, prev_obs, prev_time, first_obs)
```
Done

mgharamti commented on observations/obs_converters/BATS/bats_to_clim_obs.f90:
```diff
@@ -0,0 +1,368 @@
+   end if
+
+   ! adding the observation
+   call create_3d_obs(bats_lat, bats_lon, vert, VERTISHEIGHT, &
```
Yes, these don't change. The horizontal location is set and only the vertical position of the observations is different.

mgharamti commented:
@hkershaw-brown I incorporated all of your comments and suggestions from the obs converter.

hkershaw-brown commented:
@mgharamti do you have a wrfhydro case (or other model) you can run with this branch? 
It would be good to get a sanity check on the netcdf changes.

mgharamti commented:
@hkershaw-brown this is to confirm that the netcdf changes are good. I've ran filter with different `wrf_hydro` cases and all look good.

hkershaw-brown commented:
nice thanks @mgharamti

hkershaw-brown commented:
@mgharamti this is good to go, but let's release it next week (paranoia for netcdf changes on a Friday)

mgharamti commented:
sounds good to me. thanks for your help reviewing this.
